File: Maximize the Number of Partitions After Operations.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int maxPartitionsAfterOperations(String S, int K) {
        int N = S.length();
        int[] A = new int[N];
        for (int i = 0; i < N; i++) {
            A[i] = 1 << (S.charAt(i) - 'a');
        }
        Map<String, Integer> cache = new HashMap<>();
        return dp(0, 1, 0, A, K, cache) + 1;
    }
    
    private int dp(int i, int option, int mask, int[] A, int K, Map<String, Integer> cache) {
        if (i == A.length) {
            return 0;
        }
        String key = i + "_" + option + "_" + mask;
        if (cache.containsKey(key)) {
            return cache.get(key);
        }
        int mask2 = mask | A[i];
        int ans;
        if (Integer.bitCount(mask2) > K) {
            ans = 1 + dp(i + 1, option, A[i], A, K, cache);
        } else {
            ans = dp(i + 1, option, mask2, A, K, cache);
        }
        if (option == 1) {
            for (int j = 0; j < 26; j++) {
                mask2 = mask | (1 << j);
                if (Integer.bitCount(mask2) > K) {
                    ans = Math.max(ans, 1 + dp(i + 1, 0, 1 << j, A, K, cache));
                } else {
                    ans = Math.max(ans, dp(i + 1, 0, mask2, A, K, cache));
                }
            }
        }
        cache.put(key, ans);
        return ans;
    }
}

File: Maximize the Number of Partitions After Operations.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int maxPartitionsAfterOperations(string S, int K) {
        int N = S.length();
        vector<int> A(N);
        for (int i = 0; i < N; i++) {
            A[i] = 1 << (S[i] - 'a');
        }
        unordered_map<int, int> cache;
        function<int(int, bool, int)> dp = [&](int i, bool option, int mask) {
            if (i == N) {
                return 0;
            }
            int mask2 = mask | A[i];
            if (__builtin_popcount(mask2) > K) {
                int ans = 1 + dp(i + 1, option, A[i]);
                return ans;
            } else {
                int ans = dp(i + 1, option, mask2);
                if (option) {
                    for (int j = 0; j < 26; j++) {
                        mask2 = mask | (1 << j);
                        if (__builtin_popcount(mask2) > K) {
                            ans = max(ans, 1 + dp(i + 1, false, 1 << j));
                        } else {
                            ans = max(ans, dp(i + 1, false, mask2));
                        }
                    }
                }
                return ans;
            }
        };
        return dp(0, true, 0) + 1;
    }
};

int main() {
    Solution solution;
    string S = "abcde";
    int K = 2;
    int result = solution.maxPartitionsAfterOperations(S, K);
    cout << result << endl;
    return 0;
}


